go/ingest/ws_csv.go

```diff
@@ -0,0 +1,233 @@
+package ingest
+
+import (
+	"bytes"
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/estuary/flow/go/flow"
+	pf "github.com/estuary/flow/go/protocol"
+)
+
+type fieldMapping struct {
+	parse         []func(string) (interface{}, error)
+	field         string
+	ptr           string
+	isOptional    bool
+	possibleTypes []string
+}
+
+func (self *fieldMapping) extractValue(input string) (interface{}, error) {
+	// Is this value allowed to be empty? If it is, then we'll let the parsers determine the proper
+	// value to return.
+	if len(input) == 0 && !self.isOptional {
+		return nil, fmt.Errorf("value cannot be null")
+	}
+	var val interface{}
+	var lastErr error
+	for _, parser := range self.parse {
+		val, lastErr = parser(input)
+		if lastErr == nil {
+			return val, nil
+		}
+	}
+	return nil, fmt.Errorf("failed to parse '%v' (of column: %v) into %v: %w", input, self.field, self.possibleTypes, lastErr)
+}
+
+func hasType(projection *pf.Projection, typeName string) bool {
+	for _, ty := range projection.Inference.Types {
+		if ty == typeName {
+			return true
+		}
+	}
+	return false
+}
+
+// We traverse these types in a fairly specific order, which determines the order in which parsers
+// are applied in the case that a field may have multiple types. The first type that parses
+// successfully will be used.
+var orderedTypes = []string{"integer", "number", "boolean", "null", "string", "object", "array"}
+
+func newFieldMapping(projection *pf.Projection) fieldMapping {
+	// Whether it's valid for this field to have an empty value as input. This will be true if the
+	// allowable types include either null or string
+	var isOptional = false
+	var parse []func(string) (interface{}, error)
+	for _, typeName := range orderedTypes {
+		if !hasType(projection, typeName) {
+			continue
+		}
+		switch typeName {
+		case "null":
+			// empty input will be interpreted as null. This will take precedence over empty strings
+			isOptional = true
+			parse = append(parse, parseNull)
+		case "string":
+			// empty input will be interpreted as an empty string
+			isOptional = true
+			parse = append(parse, parseString)
+		case "boolean":
+			parse = append(parse, parseBoolean)
+		case "integer":
+			parse = append(parse, parseInt)
+		case "number":
+			parse = append(parse, parseNumber)
+		case "object":
+			parse = append(parse, parseObject)
+		case "array":
+			parse = append(parse, parseArray)
+		}
+	}
+	return fieldMapping{
+		field:         projection.Field,
+		ptr:           projection.Ptr,
+		parse:         parse,
+		isOptional:    isOptional,
+		possibleTypes: projection.Inference.Types,
+	}
+}
+
+type wsCsvIngester struct {
+	buffer             *bytes.Buffer
+	csvReader          *csv.Reader
+	projections        []fieldMapping
+	pointers           []flow.Pointer
+	lastMustExistIndex int
+}
+
+// First frame is headers, subsequent frames are documents.
+func (self *wsCsvIngester) onHeader(collection *pf.CollectionSpec) error {
```

Comment:
can you run a linter over this? It will tell you `self` is an anti-pattern, for example.

VSCode's Go tooling should make this easy.